harness/determined/common/detignore.py

```diff
@@ -0,0 +1,99 @@
+import os
+import pathlib
+from typing import TYPE_CHECKING, Callable, Iterable, List, Set
+
+if TYPE_CHECKING:
+    import pathspec
+
+from determined.common import constants, v1file_utils
+from determined.common.api import bindings
+
+
+def _build_detignore_pathspec(root_path: pathlib.Path) -> "pathspec.PathSpec":
+    ignore = list(constants.DEFAULT_DETIGNORE)
+    ignore_path = root_path / ".detignore"
+    if ignore_path.is_file():
+        with ignore_path.open("r") as detignore_file:
+            ignore.extend(detignore_file)
+
+    # Lazy import to speed up load time.
+    # See https://github.com/determined-ai/determined/pull/6590 for details.
+    import pathspec
+
+    return pathspec.PathSpec.from_lines(pathspec.patterns.GitWildMatchPattern, ignore)
+
+
+def make_shutil_ignore(root_path: pathlib.Path) -> Callable:
+    ignore_spec = _build_detignore_pathspec(root_path)
+
+    def _ignore(path: str, names: List[str]) -> Set[str]:
+        ignored_names = set()  # type: Set[str]
+        for name in names:
+            if name == ".detignore":
+                ignored_names.add(name)
+                continue
+
+            file_path = pathlib.Path(path) / name
+            file_rel_path = file_path.relative_to(root_path)
+
+            if (
+                file_path.is_dir() and ignore_spec.match_file(str(file_rel_path) + "/")
+            ) or ignore_spec.match_file(str(file_rel_path)):
```

Comment:
nit: why bounce back and forth between pathlib.Path and str?  Why not just use os.path.join() and os.path.relpath()?

non-blocking; feel free to ignore my inner C programmer which cringes at unnecessary `malloc()` calls.